File: apps/web/lib/api/files.ts
import { apiFetch } from "./client";

export type FileContentType = "image/png" | "image/jpeg" | "image/webp" | "application/pdf";
export type FileStatus = "active" | "deleted";

export type FileMetadata = {
  id: string;
  original_filename: string;
  content_type: FileContentType;
  size_bytes: number;
  expense_id: string | null;
  uploaded_by: string;
  status: FileStatus;
  created_at: string;
  deleted_at?: string | null;
  deleted_by?: string | null;
};

export type SignedUrlResponse = {
  url: string;
  expires_in: number;
};

export type UploadFileInput = {
  file: File;
  expenseId?: string | null;
};

export async function uploadFile(workspaceId: string, input: UploadFileInput) {
  const formData = new FormData();
  formData.append("file", input.file);
  if (input.expenseId) {
    formData.append("expense_id", input.expenseId);
  }

  return apiFetch<FileMetadata>(`/workspaces/${workspaceId}/files`, {
    method: "POST",
    body: formData,
  });
}

export async function listFiles(workspaceId: string) {
  return apiFetch<{ files: FileMetadata[] }>(`/workspaces/${workspaceId}/files`);
}

export async function getFile(workspaceId: string, fileId: string) {
  return apiFetch<FileMetadata>(`/workspaces/${workspaceId}/files/${fileId}`);
}

export async function getFileDownloadUrl(workspaceId: string, fileId: string) {
  return apiFetch<SignedUrlResponse>(`/workspaces/${workspaceId}/files/${fileId}/download-url`);
}

export async function linkFileToExpense(workspaceId: string, fileId: string, expenseId: string) {
  return apiFetch<FileMetadata>(`/workspaces/${workspaceId}/files/${fileId}/link`, {
    method: "POST",
    body: JSON.stringify({ expense_id: expenseId }),
  });
}

export async function detachFileFromExpense(workspaceId: string, fileId: string) {
  return apiFetch<FileMetadata>(`/workspaces/${workspaceId}/files/${fileId}/link`, {
    method: "DELETE",
  });
}

export async function deleteFile(workspaceId: string, fileId: string) {
  return apiFetch<{ id: string; status: "deleted"; deleted_at: string }>(
    `/workspaces/${workspaceId}/files/${fileId}`,
    { method: "DELETE" },
  );
}

File: apps/api/app/schemas/files.py
"""Schemas for receipt and invoice file metadata APIs."""

from __future__ import annotations

from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


FileContentType = Literal["image/png", "image/jpeg", "image/webp", "application/pdf"]
FileStatus = Literal["active", "deleted"]
FileErrorCode = Literal[
    "unauthenticated",
    "forbidden",
    "not_found",
    "file_deleted",
    "file_too_large",
    "unsupported_file_type",
    "empty_file",
    "cross_workspace_link",
]


class FileMetadata(BaseModel):
    id: UUID
    original_filename: str
    content_type: FileContentType
    size_bytes: int
    expense_id: UUID | None = None
    uploaded_by: UUID
    status: FileStatus
    created_at: datetime
    deleted_at: datetime | None = None
    deleted_by: UUID | None = None

    model_config = ConfigDict(from_attributes=True)


class FileListResponse(BaseModel):
    files: list[FileMetadata]


class LinkRequest(BaseModel):
    expense_id: UUID

    model_config = ConfigDict(extra="forbid")


class SignedUrlResponse(BaseModel):
    url: str
    expires_in: int = Field(ge=1, le=300)


class FileErrorPayload(BaseModel):
    code: FileErrorCode
    message: str


class FileTooLargeErrorPayload(FileErrorPayload):
    code: Literal["file_too_large"] = "file_too_large"
    limit_bytes: int = Field(default=10485760)
